backend/src/halo2_impl.rs

```diff
@@ -11,19 +11,49 @@ impl<T: FieldElement> BackendImpl<T> for Halo2Prover {
         Halo2Prover::new(degree)
     }
 
+    fn add_verification_key(
+        &mut self,
+        pil: &Analyzed<T>,
+        fixed: &[(String, Vec<T>)],
+        vkey: Vec<u8>,
+    ) {
+        self.add_verification_key(pil, fixed, vkey)
+    }
+
+    fn verification_key(
+        &self,
+        pil: &Analyzed<T>,
+        fixed: &[(String, Vec<T>)],
+    ) -> Result<Vec<u8>, Error> {
+        match self.verification_key(pil, fixed) {
+            Ok(vkey) => Ok(vkey),
+            Err(e) => Err(Error::VerificationKeyFailed(e)),
+        }
+    }
+
+    fn verify(&self, proof: &Proof, instances: &[Vec<T>]) -> Result<(), Error> {
+        match self.verify(proof, instances) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(Error::VerificationFailed(e)),
+        }
+    }
+
     fn prove(
         &self,
         pil: &Analyzed<T>,
         fixed: &[(String, Vec<T>)],
         witness: &[(String, Vec<T>)],
         prev_proof: Option<Proof>,
-    ) -> (Option<Proof>, Option<String>) {
+    ) -> Result<(Option<Proof>, Option<String>), Error> {
```

Comment:
I don't think we need the first `Option` anymore. Either we have a proof or there was an error. There is no scenario there is no error and no proof.

Reply:
Done